src/tarski/syntax/terms.py

```diff
@@ -187,6 +198,167 @@ def is_syntactically_equal(self, other):
+                    msg='IfThenElse: both subterms need to be of the same sort! lhs: "{}"({}), rhs: "{}"({})'.format(
+                        subterms[0], subterms[0].sort, subterms[1], subterms[1].sort))
+        else:
+            self._sort = subterms[0].sort
+
+        self.subterms = tuple(subterms)
+
+        self.symbol.language.language_components_frozen = True
+
+    @property
+    def language(self):
+        return self.symbol.language
+
+    @property
+    def sort(self):
+        return self._sort
+
+    def __str__(self):
+        return '{}({})'.format(self.symbol, ', '.join([str(self.condition)] + [str(t) for t in self.subterms]))
+
+    __repr__ = __str__
+
+    def __hash__(self):
+        return hash(('ite', self.condition, tuple(x for x in self.subterms)))
+
+    def is_syntactically_equal(self, other):
+        if (self.__class__ is not other.__class__ or self.symbol != other.symbol \
+            or len(self.subterms) != len(other.subterms)) \
+                or (not self.condition.is_syntactically_equal(other.condition)):
+            return False
+
+        # Else we just need to recursively check if all subterms are syntactically equal
+        return all(x.is_syntactically_equal(y) for x, y in zip(self.subterms, other.subterms))
+
+
+def ite(c, t1, t2):
+    return IfThenElse(c, (t1, t2))
+
+
+class Matrix(Term):
+
+    def __init__(self, arraylike, sort: Sort):
+        self.matrix = np.matrix(arraylike, dtype=np.dtype(object))
+        self._sort = sort
+        # verify and cast
+        N, M = self.matrix.shape
+        for i in range(N):
+            for j in range(M):
+                m_ij = self.matrix[i, j]
+                if isinstance(m_ij, Term):
+                    if m_ij.sort != sort:
+                        raise err.SyntacticError("Matrix: all \
+                        entries need to be of sort '{}', \
+                        entry ({},{}) is '{}'".format(sort.name, i, j, m_ij.sort.name))
+                else:
+                    self.matrix[i, j] = Constant(sort.cast(m_ij), sort)
+
+    @property
+    def shape(self):
+        return self.matrix.shape
+
+    @property
+    def language(self):
+        return self.sort.language
+
+    @property
+    def sort(self):
+        return self._sort
+
+    def __setitem__(self, key, value):
+        self.matrix[key] = value
+
+    def __getitem__(self, arg):
+        i, j = arg
+        return self.matrix[i, j]
+
+    def __str__(self):
+        return '{}'.format(self.matrix)
+
+    __repr__ = __str__
+
+    def __hash__(self):
+        N, M = self.matrix.shape
+        syms = []
+        for i in range(N):
+            for j in range(M):
+                syms += [self.matrix[i, j]]
+        return hash(tuple(syms) + self.matrix.shape)
+
+    def is_syntactically_equal(self, other):
+        if self.__class__ is not other.__class__ or self.matrix.shape != other.matrix.shape:
+            return False
+        for i in range(N):
```

Comment:
My IDE tells me that N and M are not defined here :-)
Are we really using this somewhere, or is it an experimental feature?
Are we testing it?

Reply:
We're using the ```Matrix``` classes, but the test cases aren't covering direct comparison of matrices that goes beyond line 354 above. See commit 7e946455b61eb61ca0250770ef82002ee0ea078c for adding missing line.